src/java.base/share/classes/java/security/Provider.java

```diff
@@ -1293,7 +1293,7 @@ public Service getService(String type, String algorithm) {
         Service s = serviceMap.get(key);
         if (s == null) {
             s = legacyMap.get(key);
-            if (s != null && !s.isValid()) {
+            if (s != null && (!s.isValid() || !RestrictedSecurity.isServiceAllowed(s))) {
                 legacyMap.remove(key, s);
```

Comment:
This should set `s = null;` to avoid returning a disallowed service (even once).

I expect `null` should also be returned in the case of an invalid service, but I think all services created here are `isValid()`.